planner/core/hashcode.go

```diff
@@ -97,3 +76,266 @@ func (p *LogicalLimit) HashCode() []byte {
+	result = codec.EncodeAndSort(result, lHashCode, len(p.LeftConditions))
+
+	rHashCode := func(i int) []byte { return p.RightConditions[i].HashCode(p.ctx.GetSessionVars().StmtCtx) }
+	result = codec.EncodeAndSort(result, rHashCode, len(p.RightConditions))
+
+	otherHashCode := func(i int) []byte { return p.OtherConditions[i].HashCode(p.ctx.GetSessionVars().StmtCtx) }
+	result = codec.EncodeAndSort(result, otherHashCode, len(p.OtherConditions))
+
+	return result
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *LogicalAggregation) HashCode() []byte {
+	// PlanType + SelectOffset + AggFuncs[0].Mode + p.aggHints.preferAggType + p.aggHints.preferAggToCop + Encode(AggFuncs) + Encode(GroupByItems)
+	// AggFuncs are commonly `ScalarFunction`s, whose hashcode usually has a
+	// length larger than 20, so we pre-alloc 25 bytes for each expr's hashcode.
+	// ByItems are commonly Column which hashcode has the length 9,
+	// so we pre-alloc 10 bytes for each ByItems's hashcode.
+	result := make([]byte, 0, 25+len(p.AggFuncs)*29+len(p.GroupByItems)*14)
+	result = codec.EncodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = codec.EncodeIntAsUint32(result, p.SelectBlockOffset())
+	result = codec.EncodeIntAsUint32(result, int(p.AggFuncs[0].Mode))
+
+	result = codec.EncodeUintAsUint32(result, uint(p.aggHints.preferAggType))
+	result = codec.EncodeBool(result, p.aggHints.preferAggToCop)
+
+	aggFuncHashCode := func(i int) []byte { return p.AggFuncs[i].HashCode(p.ctx.GetSessionVars().StmtCtx) }
+	result = codec.Encode(result, aggFuncHashCode, len(p.AggFuncs))
+
+	groupByHashCode := func(i int) []byte { return p.GroupByItems[i].HashCode(p.ctx.GetSessionVars().StmtCtx) }
+	result = codec.EncodeAndSort(result, groupByHashCode, len(p.GroupByItems))
+
+	return result
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *LogicalMaxOneRow) HashCode() []byte {
+	// PlanType + SelectOffset
+	result := make([]byte, 0, 8)
+	result = codec.EncodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = codec.EncodeIntAsUint32(result, p.SelectBlockOffset())
+	return result
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *LogicalShowDDLJobs) HashCode() []byte {
+	// PlanType + SelectOffset + JobNumber
+	result := make([]byte, 16)
+	result = codec.EncodeIntAsUint32(result, plancodec.TypeStringToPhysicalID(p.tp))
+	result = codec.EncodeIntAsUint32(result, p.SelectBlockOffset())
+	binary.BigEndian.PutUint64(result[8:], uint64(p.JobNumber))
+	return result
+}
+
+// HashCode implements LogicalPlan interface.
+func (p *DataSource) HashCode() []byte {
+	// PlanType + SelectOffset + id + isPartition + physicalTableID
+	// usually we would not copy a dataSource same as other one and we can use PlanID as hashCode.
+	// but in rule_partition_processor, it will copy the old dataSource,
+	// and only change isPartition,physicalTableID and share the same PlanID.
+	// So we should append isPartition and physicalTableID.
```

Comment:
`HashCode` will only be used for cascades planner.  We don't need consider other rules in plannercore. IMHO, `DataSource` only needs to encode information which could indentify the `Table`, and the pushed-down conditions.